lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO/Gen.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2024 Cardano Foundation
+-- License: Apache-2.0
+--
+-- Provides generators and shrinkers for the 'TxWithUTxO' data type.
+--
+module Internal.Cardano.Write.Tx.TxWithUTxO.Gen
+    ( generate
+    , generateWithMinimalUTxO
+    , generateWithSurplusUTxO
+    , shrink
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Api
+    ( EraTx (bodyTxL)
+    )
+import Cardano.Ledger.Api.Tx.Body
+    ( allInputsTxBodyF
+    )
+import Control.Lens
+    ( view
+    )
+import Internal.Cardano.Write.Tx
+    ( IsRecentEra
+    , Tx
+    , TxIn
+    , TxOut
+    , UTxO (UTxO, unUTxO)
+    )
+import Internal.Cardano.Write.Tx.TxWithUTxO
+    ( pattern TxWithUTxO
+    , type TxWithUTxO
+    )
+import Test.QuickCheck
+    ( Gen
+    , oneof
+    )
+import Test.QuickCheck.Extra
+    ( genMapFromKeysWith
+    , genMapWith
+    , shrinkMapToSubmaps
+    )
+
+import qualified Data.Map.Strict as Map
+import qualified Internal.Cardano.Write.Tx.TxWithUTxO as TxWithUTxO
+
+-- | Generates a 'TxWithUTxO' object.
+--
+-- The domain of the UTxO map is a superset of the transaction input set, but
+-- it may or may not be a strict superset.
+--
+generate
+    :: IsRecentEra era
+    => Gen (Tx era)
+    -> Gen (TxIn)
+    -> Gen (TxOut era)
+    -> Gen (TxWithUTxO era)
+generate genTx genTxIn genTxOut =
+    oneof
+        [ generateWithMinimalUTxO genTx genTxIn genTxOut
+        , generateWithSurplusUTxO genTx genTxIn genTxOut
+        ]
+
+-- | Generates a 'TxWithUTxO' object that has a minimal UTxO set.
+--
+-- The domain of the UTxO map is exactly equal to the transaction input set.
+--
+generateWithMinimalUTxO
+    :: IsRecentEra era
+    => Gen (Tx era)
+    -> Gen (TxIn)
+    -> Gen (TxOut era)
+    -> Gen (TxWithUTxO era)
+generateWithMinimalUTxO genTx _genTxIn genTxOut = do
+    tx <- genTx
+    utxo <- UTxO <$> genMapFromKeysWith genTxOut (txInputs tx)
+    pure $ TxWithUTxO.constructFiltered tx utxo
+  where
+    txInputs = view (bodyTxL . allInputsTxBodyF)
+
+-- | Generates a 'TxWithUTxO' object that has a non-minimal UTxO set.
+--
+-- The domain of the UTxO map is a strict superset of the transaction input set.
```

Comment:
```suggestion
-- The domain of the UTxO map is a superset of the transaction input set.
```

Couldn't `surplusUTxO` be a subset of `minimalUTxO` such that `minimalUTxO <> surplusUTxO == minimalUTxO`?

Reply:
@Anviking I've changed the generator to guarantee that the result is a strict superset.